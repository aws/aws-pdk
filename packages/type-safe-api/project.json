{
  "name": "@aws/type-safe-api",
  "root": "packages/type-safe-api",
  "targets": {
    "build": {
      "inputs": [
        "default",
        "^default",
        "!{projectRoot}/.jsii",
        "!{projectRoot}/lib/**/*",
        "!{projectRoot}/dist/**/*",
        "!{projectRoot}/coverage/**/*",
        "!{projectRoot}/test-reports/**/*",
        "!{projectRoot}/docs/api/**/*"
      ],
      "outputs": [
        "{projectRoot}/.jsii",
        "{projectRoot}/lib",
        "{projectRoot}/dist",
        "{projectRoot}/coverage",
        "{projectRoot}/test-reports",
        "{projectRoot}/docs/api"
      ],
      "dependsOn": [
        "^build"
      ],
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen build",
        "cwd": "packages/type-safe-api"
      }
    },
    "default": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen default",
        "cwd": "packages/type-safe-api"
      }
    },
    "pre-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen pre-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "post-compile": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen post-compile",
        "cwd": "packages/type-safe-api"
      }
    },
    "test": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test",
        "cwd": "packages/type-safe-api"
      }
    },
    "package": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package",
        "cwd": "packages/type-safe-api"
      }
    },
    "test:watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen test:watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "watch": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen watch",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint",
        "cwd": "packages/type-safe-api"
      }
    },
    "compat": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen compat",
        "cwd": "packages/type-safe-api"
      }
    },
    "package-all": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package-all",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:java": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:java",
        "cwd": "packages/type-safe-api"
      }
    },
    "package:python": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen package:python",
        "cwd": "packages/type-safe-api"
      }
    },
    "eslint-staged": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen eslint-staged",
        "cwd": "packages/type-safe-api"
      }
    },
    "jest": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen jest",
        "cwd": "packages/type-safe-api"
      }
    },
    "docgen": {
      "executor": "nx:run-commands",
      "options": {
        "command": "pnpm exec projen docgen",
        "cwd": "packages/type-safe-api"
      }
    }
  },
  "implicitDependencies": [
    "smithy-async-transformer"
  ],
  "//": "~~ Generated by projen. To modify, edit .projenrc.js and run \"pnpm exec projen\"."
}
