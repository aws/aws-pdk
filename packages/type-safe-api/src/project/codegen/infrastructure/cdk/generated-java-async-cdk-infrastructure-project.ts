/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import {
  GeneratedJavaCdkInfrastructureBaseProject,
  GeneratedJavaCdkInfrastructureBaseProjectOptions,
} from "./generated-java-cdk-infrastructure-base-project";
import {
  GenerationOptions,
  OtherGenerators,
  getHandlersProjectVendorExtensions,
} from "../../components/utils";

export interface GeneratedJavaAsyncCdkInfrastructureProjectOptions
  extends GeneratedJavaCdkInfrastructureBaseProjectOptions {}

export class GeneratedJavaAsyncCdkInfrastructureProject extends GeneratedJavaCdkInfrastructureBaseProject {
  constructor(options: GeneratedJavaAsyncCdkInfrastructureProjectOptions) {
    super(options);
  }

  public buildOpenApiGeneratorOptions(): GenerationOptions {
    return {
      generator: "java",
      specPath: this.options.specPath,
      generatorDirectory: OtherGenerators.JAVA_ASYNC_CDK_INFRASTRUCTURE,
      srcDir: this.srcDir,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
      extraVendorExtensions: {
        "x-infrastructure-package": this.packageName,
        "x-runtime-package": this.options.generatedJavaTypes.packageName,
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
      // Do not generate map/list types. Generator will use built in HashMap, ArrayList instead
      generateAliasAsModel: false,
    };
  }
}
