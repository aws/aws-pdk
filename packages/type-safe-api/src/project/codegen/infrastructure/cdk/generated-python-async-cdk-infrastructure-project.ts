/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import {
  GeneratedPythonCdkInfrastructureBaseProject,
  GeneratedPythonCdkInfrastructureBaseProjectOptions,
} from "./generated-python-cdk-infrastructure-base-project";
import {
  CodegenOptions,
  OtherGenerators,
  getHandlersProjectVendorExtensions,
} from "../../components/utils";

export interface GeneratedPythonAsyncCdkInfrastructureProjectOptions
  extends GeneratedPythonCdkInfrastructureBaseProjectOptions {}

export class GeneratedPythonAsyncCdkInfrastructureProject extends GeneratedPythonCdkInfrastructureBaseProject {
  constructor(options: GeneratedPythonAsyncCdkInfrastructureProjectOptions) {
    super(options);
  }

  public buildCodegenOptions(): CodegenOptions {
    return {
      specPath: this.options.specPath,
      templateDirs: [OtherGenerators.PYTHON_ASYNC_CDK_INFRASTRUCTURE],
      metadata: {
        srcDir: this.moduleName,
        runtimeModuleName: this.options.generatedPythonTypes.moduleName,
        relativeSpecPath: path.join("..", this.options.specPath),
        ...getHandlersProjectVendorExtensions(
          this,
          this.options.generatedHandlers
        ),
      },
    };
  }
}
