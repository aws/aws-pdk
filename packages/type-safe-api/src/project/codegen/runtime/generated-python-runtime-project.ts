/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import { PythonProject } from "projen/lib/python";
import { Language } from "../../languages";
import {
  CodeGenerationSourceOptions,
  GeneratedPythonRuntimeOptions,
} from "../../types";
import { OpenApiGeneratorIgnoreFile } from "../components/open-api-generator-ignore-file";
import { OpenApiToolsJsonFile } from "../components/open-api-tools-json-file";
import {
  buildCleanOpenApiGeneratedCodeCommand,
  buildInvokeOpenApiGeneratorCommandArgs,
  buildTypeSafeApiExecCommand,
  TypeSafeApiScript,
} from "../components/utils";

/**
 * Configuration for the generated python types project
 */
export interface GeneratedPythonTypesProjectOptions
  extends GeneratedPythonRuntimeOptions,
    CodeGenerationSourceOptions {}

/**
 * Python project containing types generated using OpenAPI Generator CLI
 */
export class GeneratedPythonRuntimeProject extends PythonProject {
  /**
   * Patterns that are excluded from code generation
   */
  public static openApiIgnorePatterns: string[] = [
    "test",
    "test/*",
    "test/**/*",
    ".github",
    ".github/*",
    ".github/**/*",
    ".gitlab-ci.yml",
    ".travis.yml",
    "git_push.sh",
    "tox.ini",
    "requirements.txt",
    "test-requirements.txt",
    "setup.py",
    "setup.cfg",
  ];

  /**
   * Options configured for the project
   * @private
   */
  private readonly options: GeneratedPythonTypesProjectOptions;

  constructor(options: GeneratedPythonTypesProjectOptions) {
    super({
      ...(options as any),
      sample: false,
      pytest: false,
      poetry: true,
      poetryOptions: {
        packages: [{ include: options.moduleName }],
        // Module must be explicitly added to include since poetry excludes everything in .gitignore by default
        include: [options.moduleName, `${options.moduleName}/**/*.py`],
      },
    });
    this.options = options;

    // Add dependencies required by the client
    [
      "python-dateutil@~2.8.2",
      "pydantic@^1.10.5",
      "aenum@^3.1.11",
      "urllib3@~1.26.7",
      `aws-lambda-powertools@{extras=["all"],version="^2.23.0"}`,
      "python@^3.9",
    ].forEach((dep) => this.addDependency(dep));

    const openapiGeneratorIgnore = new OpenApiGeneratorIgnoreFile(this);
    openapiGeneratorIgnore.addPatterns(
      ...GeneratedPythonRuntimeProject.openApiIgnorePatterns
    );

    // Add OpenAPI Generator cli configuration
    OpenApiToolsJsonFile.ensure(this).addOpenApiGeneratorCliConfig({
      version: "6.6.0",
      ...options.openApiGeneratorCliConfig,
    });

    const generateTask = this.addTask("generate");
    generateTask.exec(buildCleanOpenApiGeneratedCodeCommand());
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE,
        this.buildGenerateCommandArgs()
      )
    );

    this.preCompileTask.spawn(generateTask);

    // Ignore all the generated code
    this.gitignore.addPatterns(
      this.moduleName,
      "docs",
      "README.md",
      ".openapi-generator"
    );

    // The poetry install that runs as part of post synthesis expects there to be some code present, but code isn't
    // generated until build time. This means that the first install will fail when either generating the project for
    // the first time or checking out a fresh copy (since generated code is not checked in to version control). We
    // therefore add a blank __init__.py as our first install step to keep poetry happy until the generator overwrites
    // it.
    this.tasks
      .tryFind("install")
      ?.prependExec(
        `mkdir -p ${this.moduleName} && touch ${this.moduleName}/__init__.py`
      );
  }

  public buildGenerateCommandArgs = () => {
    return buildInvokeOpenApiGeneratorCommandArgs({
      generator: "python-nextgen",
      specPath: this.options.specPath,
      smithyJsonPath: this.options.smithyJsonModelPath,
      generatorDirectory: Language.PYTHON,
      additionalProperties: {
        packageName: this.moduleName,
        projectName: this.name,
      },
      // Tell the generator where python source files live
      srcDir: this.moduleName,
      normalizers: {
        KEEP_ONLY_FIRST_TAG_IN_OPERATION: true,
      },
    });
  };
}
