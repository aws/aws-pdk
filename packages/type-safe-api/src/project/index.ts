/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
export * from "./model/type-safe-api-model-build";
export * from "./model/type-safe-api-async-model-build";
export * from "./model/smithy/smithy-project-definition";
export * from "./model/openapi/open-api-project-definition";
export * from "./type-safe-api-project";
export * from "./model/type-safe-api-model-project";
export * from "./model/smithy/smithy-definition";
export * from "./model/openapi/open-api-definition";
export * from "./type-safe-websocket-api-project";
export * from "./model/type-safe-websocket-api-model-project";
export * from "./model/smithy/smithy-async-definition";
export * from "./model/openapi/open-api-async-definition";
export * from "./model/smithy/types";
export * from "./model/openapi/open-api-model-project";
export * from "./model/openapi/open-api-async-model-project";
export * from "./model/smithy/smithy-async-model-project";
export * from "./model/smithy/smithy-model-project";
export * from "./languages";
export * from "./types";
export * from "./typescript-project-options";
export * from "./python-project-options";
export * from "./java-project-options";
