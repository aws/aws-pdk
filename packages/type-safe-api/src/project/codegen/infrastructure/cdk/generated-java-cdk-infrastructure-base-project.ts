/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { ProjectUtils } from "@aws/monorepo";
import { DependencyType, Task } from "projen";
import { JavaProject, JavaProjectOptions } from "projen/lib/java";
import {
  CodeGenerationSourceOptions,
  GeneratedWithOpenApiGeneratorOptions,
} from "../../../types";
import { TypeSafeApiCommandEnvironment } from "../../components/type-safe-api-command-environment";
import {
  buildCodegenCommandArgs,
  buildTypeSafeApiExecCommand,
  CodegenOptions,
  TypeSafeApiScript,
} from "../../components/utils";
import { GeneratedHandlersProjects } from "../../generate";
import { GeneratedJavaRuntimeBaseProject } from "../../runtime/generated-java-runtime-base-project";

export interface GeneratedJavaInfrastructureBaseOptions
  extends JavaProjectOptions,
    GeneratedWithOpenApiGeneratorOptions {}

export interface GeneratedJavaCdkInfrastructureBaseProjectOptions
  extends GeneratedJavaInfrastructureBaseOptions,
    CodeGenerationSourceOptions {
  /**
   * The generated java types
   */
  readonly generatedJavaTypes: GeneratedJavaRuntimeBaseProject;
  /**
   * Generated handlers projects
   */
  readonly generatedHandlers: GeneratedHandlersProjects;
}

export abstract class GeneratedJavaCdkInfrastructureBaseProject extends JavaProject {
  /**
   * Options configured for the project
   */
  protected readonly options: GeneratedJavaCdkInfrastructureBaseProjectOptions;

  /**
   * Source directory
   */
  protected readonly srcDir: string;

  /**
   * Java package name
   */
  protected readonly packageName: string;

  /**
   * Code generation task
   */
  protected readonly generateTask: Task;

  constructor(options: GeneratedJavaCdkInfrastructureBaseProjectOptions) {
    super({
      ...options,
      sample: false,
      junit: false,
    });
    TypeSafeApiCommandEnvironment.ensure(this);

    this.options = options;
    this.packageName = `${this.pom.groupId}.${this.name}.infra`;
    this.srcDir = path.join(
      "src",
      "main",
      "java",
      ...this.packageName.split(".")
    );

    [
      `software.aws/pdk@${ProjectUtils.getPdkVersion()}`,
      "software.constructs/constructs@10.3.0",
      "software.amazon.awscdk/aws-cdk-lib@2.133.0",
      "io.github.cdklabs/cdknag@2.28.60",
      "org.projectlombok/lombok@1.18.30",
      "com.fasterxml.jackson.core/jackson-databind@2.17.0",
      `io.github.cdklabs/projen@0.80.10`,
      `${options.generatedJavaTypes.pom.groupId}/${options.generatedJavaTypes.pom.artifactId}@${options.generatedJavaTypes.pom.version}`,
    ]
      .filter(
        (dep) =>
          !this.deps.tryGetDependency(dep.split("@")[0], DependencyType.RUNTIME)
      )
      .forEach((dep) => this.addDependency(dep));

    // Pin constructs version
    this.deps.removeDependency(
      "software.constructs/constructs",
      DependencyType.BUILD
    );
    this.addDependency("software.constructs/constructs@10.3.0");

    // Remove the projen test dependency since otherwise it takes precedence, causing projen to be unavailable at synth time
    this.deps.removeDependency("io.github.cdklabs/projen", DependencyType.TEST);

    // Add a dependency on the generated java types repository
    this.pom.addRepository({
      url: `file://${path.relative(
        this.outdir,
        options.generatedJavaTypes.outdir
      )}/dist/java`,
      id: `${options.generatedJavaTypes.pom.groupId}-${options.generatedJavaTypes.pom.artifactId}-repo`,
    });

    const generateTask = this.addTask("generate");
    this.generateTask = generateTask;
    generateTask.exec(
      buildTypeSafeApiExecCommand(
        TypeSafeApiScript.GENERATE_NEXT,
        this.buildGenerateCommandArgs()
      )
    );
    // Copy the parsed spec into the resources directory so that it's included in the jar
    generateTask.exec("mkdir -p src/main/resources");
    generateTask.exec(
      `cp -f ${this.options.specPath} src/main/resources/.api.json`
    );
    // Absolute path of this project is required for determining the path to the handlers jar,
    // since java executes from the jar which could be anywhere in the filesystem (eg the .m2 directory).
    // While ugly, since this is written at build time and not checked in it remains portable.
    generateTask.exec(
      "echo $(pwd) > src/main/resources/project-absolute-path.txt"
    );

    this.gitignore.addPatterns(".tsapi-manifest", ".openapi-generator");

    this.preCompileTask.spawn(generateTask);
  }

  public buildGenerateCommandArgs = () => {
    return buildCodegenCommandArgs(this.buildCodegenOptions());
  };

  protected abstract buildCodegenOptions(): CodegenOptions;
}
