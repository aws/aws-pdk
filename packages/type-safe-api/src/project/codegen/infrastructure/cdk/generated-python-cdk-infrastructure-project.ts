/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import * as path from "path";
import { DependencyType } from "projen";
import { PythonProject, PythonProjectOptions } from "projen/lib/python";
import { Language } from "../../../languages";
import { buildGenerateCdkInfrastructureCommand } from "../../components/utils";
import { GeneratedPythonRuntimeProject } from "../../runtime/generated-python-runtime-project";

export interface GeneratedPythonCdkInfrastructureProjectOptions
  extends PythonProjectOptions {
  /**
   * OpenAPI spec path, relative to the project outdir
   */
  readonly specPath: string;
  /**
   * The generated python types
   */
  readonly generatedPythonTypes: GeneratedPythonRuntimeProject;
}

export class GeneratedPythonCdkInfrastructureProject extends PythonProject {
  /**
   * Path to the openapi specification
   * @private
   */
  private readonly specPath: string;

  /**
   * The generated python types
   * @private
   */
  private readonly generatedPythonTypes: GeneratedPythonRuntimeProject;

  constructor(options: GeneratedPythonCdkInfrastructureProjectOptions) {
    super({
      sample: false,
      pytest: false,
      poetry: true,
      poetryOptions: {
        packages: [{ include: options.moduleName }],
        // Module must be explicitly added to include since poetry excludes everything in .gitignore by default
        include: [options.moduleName, `${options.moduleName}/**/*.py`],
      },
      ...options,
    });
    this.specPath = options.specPath;
    this.generatedPythonTypes = options.generatedPythonTypes;

    [
      "aws_prototyping_sdk.type_safe_api@^0.0.0",
      "constructs@^10.0.0",
      "aws-cdk-lib@^2.0.0",
      "cdk-nag@^2.0.0",
      `${options.generatedPythonTypes.name}@{path="${path.relative(
        this.outdir,
        options.generatedPythonTypes.outdir
      )}", develop=true}`,
    ]
      .filter((dep) => !this.deps.tryGetDependency(dep, DependencyType.RUNTIME))
      .forEach((dep) => this.addDependency(dep));

    const generateInfraCommand = this.buildGenerateCommand();

    const generateTask = this.addTask("generate");
    generateTask.exec(generateInfraCommand.command, {
      cwd: path.relative(this.outdir, generateInfraCommand.workingDir),
    });

    this.preCompileTask.spawn(generateTask);

    // Ignore the generated code
    this.gitignore.addPatterns(this.moduleName);
  }

  public buildGenerateCommand = () => {
    return buildGenerateCdkInfrastructureCommand({
      language: Language.PYTHON,
      sourcePath: path.join(this.outdir, this.moduleName),
      generatedTypesPackage: this.generatedPythonTypes.moduleName,
      infraPackage: this.moduleName,
      // Spec path relative to the source directory
      specPath: path.join("..", this.specPath),
    });
  };
}
