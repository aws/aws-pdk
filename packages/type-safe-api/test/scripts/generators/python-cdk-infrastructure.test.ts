/*! Copyright [Amazon.com](http://amazon.com/), Inc. or its affiliates. All Rights Reserved.
SPDX-License-Identifier: Apache-2.0 */
import os from "os";
import * as path from "path";
import { exec } from "projen/lib/util";
import { GeneratedPythonCdkInfrastructureProject } from "../../../src/project/codegen/infrastructure/cdk/generated-python-cdk-infrastructure-project";
import { GeneratedPythonRuntimeProject } from "../../../src/project/codegen/runtime/generated-python-runtime-project";
import { withTmpDirSnapshot } from "../../project/snapshot-utils";

describe("Python Infrastructure Code Generation Script Unit Tests", () => {
  it.each(["single.yaml"])("Generates With %s", (spec) => {
    const specPath = path.resolve(__dirname, `../../resources/specs/${spec}`);

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedPythonRuntimeProject({
        name: "test-client",
        moduleName: "test_client",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedPythonCdkInfrastructureProject({
        name: "test-infra",
        moduleName: "test_infra",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedPythonTypes: client,
      });
      project.synth();
      exec(
        `${path.resolve(
          __dirname,
          "../../../scripts/generators/generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/test_infra/api.py"]).toMatchSnapshot();
    expect(snapshot["infra/test_infra/__init__.py"]).toMatchSnapshot();
    expect(snapshot["infra/test_infra/mock_integrations.py"]).toMatchSnapshot();
  });

  it("Generates With Mocks Disabled", () => {
    const specPath = path.resolve(
      __dirname,
      `../../resources/specs/single.yaml`
    );

    const snapshot = withTmpDirSnapshot(os.tmpdir(), (outdir) => {
      exec(`cp ${specPath} ${outdir}/spec.yaml`, {
        cwd: path.resolve(__dirname),
      });
      const clientOutdir = path.join(outdir, "client");
      const client = new GeneratedPythonRuntimeProject({
        name: "test-client",
        moduleName: "test_client",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: clientOutdir,
        specPath: "../spec.yaml",
      });
      const infraOutdir = path.join(outdir, "infra");
      const project = new GeneratedPythonCdkInfrastructureProject({
        name: "test-infra",
        moduleName: "test_infra",
        authorEmail: "me@example.com",
        authorName: "test",
        version: "1.0.0",
        outdir: infraOutdir,
        specPath: "../spec.yaml",
        generatedPythonTypes: client,
        mockDataOptions: {
          disable: true,
        },
      });
      project.synth();
      exec(
        `${path.resolve(
          __dirname,
          "../../../scripts/generators/generate"
        )} ${project.buildGenerateCommandArgs()}`,
        {
          cwd: infraOutdir,
        }
      );
    });

    expect(snapshot["infra/test_infra/mock_integrations.py"]).toMatchSnapshot();
  });
});
