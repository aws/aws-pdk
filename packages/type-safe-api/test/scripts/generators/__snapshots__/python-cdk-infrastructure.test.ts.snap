// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Python Infrastructure Code Generation Script Unit Tests Generates Functions 1`] = `
"from aws_cdk import Duration
from aws_cdk.aws_lambda import (
    Function, Runtime, Tracing, Code
)
from os import path
from pathlib import Path



class JavaOneFunction(Function):
    """
    Lambda function construct which points to the java implementation of JavaOne
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.JAVA_17,
            handler="test.test-java-handlers.handlers.JavaOneHandler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class JavaTwoFunction(Function):
    """
    Lambda function construct which points to the java implementation of JavaTwo
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.JAVA_17,
            handler="test.test-java-handlers.handlers.JavaTwoHandler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../java-handlers/dist/java/test/com.aws.pdk.test.handlers/1.0.0/com.aws.pdk.test.handlers-1.0.0.jar",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class PythonOneFunction(Function):
    """
    Lambda function construct which points to the python implementation of PythonOne
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.PYTHON_3_11,
            handler="test_python_handlers.python_one.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../python-handlers/dist/lambda",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class PythonTwoFunction(Function):
    """
    Lambda function construct which points to the python implementation of PythonTwo
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.PYTHON_3_11,
            handler="test_python_handlers.python_two.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../python-handlers/dist/lambda",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class TypescriptOneFunction(Function):
    """
    Lambda function construct which points to the typescript implementation of TypescriptOne
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.NODEJS_18_X,
            handler="index.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../typescript-handlers/dist/lambda",
                "typescript-one",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )

class TypescriptTwoFunction(Function):
    """
    Lambda function construct which points to the typescript implementation of TypescriptTwo
    """
    def __init__(self, scope, id, **kwargs):
        super().__init__(scope, id,
            runtime=Runtime.NODEJS_18_X,
            handler="index.handler",
            code=Code.from_asset(path.join(str(Path(__file__).absolute().parent), "..",
                "../typescript-handlers/dist/lambda",
                "typescript-two",
            )),
            tracing=Tracing.ACTIVE,
            timeout=Duration.seconds(30),
            **kwargs,
        )"
`;

exports[`Python Infrastructure Code Generation Script Unit Tests Generates With Mocks Disabled 1`] = `
"import json
from aws.pdk.type_safe_api import Integrations, MockIntegration, TypeSafeApiIntegration
from test_client.models import *
from test_client.api.operation_config import OperationConfig
from os import path
from pathlib import Path

MOCK_DATA_PATH = path.join(str(Path(__file__).absolute().parent), "..", "mocks")

class MockIntegrations:
    """
    Type-safe mock integrations for API operations
    """
    
    
    # No mock integrations have been generated, since mock data generation is disabled.
    pass
    
"
`;

exports[`Python Infrastructure Code Generation Script Unit Tests Generates With single.yaml 1`] = `
"from dataclasses import fields
from aws.pdk.type_safe_api import TypeSafeRestApi, TypeSafeApiIntegration
from test_client.api.operation_config import OperationLookup, OperationConfig
from os import path
from pathlib import Path

SPEC_PATH = path.join(str(Path(__file__).absolute().parent), "../../spec.yaml")

class Api(TypeSafeRestApi):
    """
    Type-safe construct for the API Gateway resources defined by your model.
    This construct is generated and should not be modified.
    """
    def __init__(self, scope, id, integrations: OperationConfig[TypeSafeApiIntegration], **kwargs):
        super().__init__(scope, id,
            **kwargs,
            integrations={ field.name: getattr(integrations, field.name) for field in fields(integrations) },
            spec_path=SPEC_PATH,
            operation_lookup=OperationLookup,
        )
"
`;

exports[`Python Infrastructure Code Generation Script Unit Tests Generates With single.yaml 2`] = `"#"`;

exports[`Python Infrastructure Code Generation Script Unit Tests Generates With single.yaml 3`] = `
"import json
from aws.pdk.type_safe_api import Integrations, MockIntegration, TypeSafeApiIntegration
from test_client.models import *
from test_client.api.operation_config import OperationConfig
from os import path
from pathlib import Path

MOCK_DATA_PATH = path.join(str(Path(__file__).absolute().parent), "..", "mocks")

class MockIntegrations:
    """
    Type-safe mock integrations for API operations
    """
    
    @staticmethod
    def any_request_response_200(body: str) -> MockIntegration:
        """
        Mock integration to return a 200 response from the any_request_response operation
        """
        return Integrations.mock(
            status_code=200,
            body=body,
        )

    @staticmethod
    def empty_204() -> MockIntegration:
        """
        Mock integration to return a 204 response from the empty operation
        """
        return Integrations.mock(
            status_code=204,
        )

    @staticmethod
    def map_response_200(body: MapResponse = None) -> MockIntegration:
        """
        Mock integration to return a 200 response from the map_response operation
        Call this with no arguments to use a generated mock response
        """
        response_body = None
        if body is None:
            with open(path.join(MOCK_DATA_PATH, "GET{}-200.json".format("/map-response".replace("/", "-")).lower()), "r") as f:
                response_body = f.read()
        else:
            response_body = body.to_json()
        return Integrations.mock(
            status_code=200,
            body=response_body,
        )

    @staticmethod
    def media_types_200(body: str) -> MockIntegration:
        """
        Mock integration to return a 200 response from the media_types operation
        """
        return Integrations.mock(
            status_code=200,
            body=body,
        )

    @staticmethod
    def multiple_content_types_200(body: str) -> MockIntegration:
        """
        Mock integration to return a 200 response from the multiple_content_types operation
        """
        return Integrations.mock(
            status_code=200,
            body=body,
        )

    @staticmethod
    def operation_one_200(body: TestResponse = None) -> MockIntegration:
        """
        Mock integration to return a 200 response from the operation_one operation
        Call this with no arguments to use a generated mock response
        """
        response_body = None
        if body is None:
            with open(path.join(MOCK_DATA_PATH, "POST{}-200.json".format("/path/{pathParam}".replace("/", "-")).lower()), "r") as f:
                response_body = f.read()
        else:
            response_body = body.to_json()
        return Integrations.mock(
            status_code=200,
            body=response_body,
        )

    @staticmethod
    def operation_one_400(body: ApiError = None) -> MockIntegration:
        """
        Mock integration to return a 400 response from the operation_one operation
        Call this with no arguments to use a generated mock response
        """
        response_body = None
        if body is None:
            with open(path.join(MOCK_DATA_PATH, "POST{}-400.json".format("/path/{pathParam}".replace("/", "-")).lower()), "r") as f:
                response_body = f.read()
        else:
            response_body = body.to_json()
        return Integrations.mock(
            status_code=400,
            body=response_body,
        )

    @staticmethod
    def without_operation_id_delete_200(body: TestResponse = None) -> MockIntegration:
        """
        Mock integration to return a 200 response from the without_operation_id_delete operation
        Call this with no arguments to use a generated mock response
        """
        response_body = None
        if body is None:
            with open(path.join(MOCK_DATA_PATH, "DELETE{}-200.json".format("/without-operation-id".replace("/", "-")).lower()), "r") as f:
                response_body = f.read()
        else:
            response_body = body.to_json()
        return Integrations.mock(
            status_code=200,
            body=response_body,
        )

    @staticmethod
    def mock_all(**kwargs) -> OperationConfig[TypeSafeApiIntegration]:
        """
        Mock all operations for which generated JSON data can be returned.
        The first available response is used for each operation. In most cases this is the successful 200 response.
        Pass any additional or overridden integrations as kwargs, for example:

        MockIntegrations.mock_all(
            say_hello=TypeSafeApiIntegration(
                integration=Integrations.lambda_(...)
            )
        )
        """
        return OperationConfig(**{
            **{
                "map_response": TypeSafeApiIntegration(
                    integration=MockIntegrations.map_response_200(),
                ),
                "operation_one": TypeSafeApiIntegration(
                    integration=MockIntegrations.operation_one_200(),
                ),
                "without_operation_id_delete": TypeSafeApiIntegration(
                    integration=MockIntegrations.without_operation_id_delete_200(),
                ),
            },
            **kwargs
        })
    
    
"
`;
